man/egalax.man
.ds q \N'34'
.TH EGALAX __drivermansuffix__ __vendorversion__
.SH NAME
egalax \- eGalax touch screen input driver
.SH SYNOPSIS
.nf
.B "Section \*qInputDevice\*q"
.BI "  Identifier \*q" idevname \*q
.B  "  Driver \*qegalax\*q"
.BI "  Option \*qDevice\*q   \*q" devpath \*q
\ \ ...
.br
.B EndSection
.SH DESCRIPTION
.B egalax 
is an __xservername__ input driver for eGalax touchscreen devices...
.PP
The
.B egalax
driver functions as a pointer input device, and may be used as the
X server's core pointer.
.PP
It is obvious that a touch panel with a stylus/finger is a one-button mouse.
However, the
.B egalax
tries to overcome this limitation.
If a user keeps stylus in one point for a certain time, then
driver emulates release of left mouse button and then pressure on
the right mouse button.
The area where the stylus should be kept and time are configurable.
.SH SUPPORTED HARDWARE
The driver has been written for CarTFT CTF1020-S on screen touch panel,
and is supposed to work with other similar touch panels based on eGalax.
.PP
The
.B egalax
driver requires the uep(4) kernel driver.
.SH CONFIGURATION DETAILS
The following driver
.B Options
are supported:
.TP 7
.BI "Option \*qDevice\*q \*q" string \*q
.RS
Usually \*q/dev/uep0\*q .
.RE
.TP 7
.BI "Option \*qMinX\*q \*q" integer \*q
.TP 7
.BI "Option \*qMaxX\*q \*q" integer \*q
.TP 7
.BI "Option \*qMinY\*q \*q" integer \*q
.TP 7
.BI "Option \*qMaxY\*q \*q" integer \*q
.RS 7
These values configure calibration of the touch panel.
.PP
The touchscreen resolution is 2048 points in both directions.
However, the onscreen part of the panel is usually smaller.
So, the panel needs to be calibrated.
Initial values for calibration are quite conservative, and
the driver autocalibrates itself when user touches areas near edge
of screen.
After each autocalibration the driver writes suggested calibration values
to the __xservername__ log.
These values should be written to the __xservername__ configuration file,
to avoid bad calibration on next run.
.RE
.TP 7
.BI "Option \*qReverseY\*q \*q" boolean \*q
Invert the Y axis.
Default: off.
.TP 7
.BI "Option \*qSwapAxes\*q \*q" boolean \*q
Some eGalax touch screen units are assembled incorrectly and have X
and Y axes swapped.
This configuration knob allows to swap them back.
Default: off.
.TP 7
.BI "Option \*qRightClickEmulArea\*q \*q" integer \*q
Size of the area where a stylus should be held still, before the driver
emulates right button click.
The default is 5 pixels.
.TP 7
.BI "Option \*qRightClickEmulPause\*q \*q" integer \*q
Time the driver waits befor emulating right button pressure.
The default is 1 second.
.PP
.SH "SEE ALSO"
__xservername__(__appmansuffix__), __xconfigfile__(__filemansuffix__), xorgconfig(__appmansuffix__), Xserver(__appmansuffix__), X(__miscmansuffix__).
.SH AUTHORS
The driver has been written by Gleb Smirnoff


src/egalax.c
/*-
 * Copyright (c) 2010, 2014 Gleb Smirnoff <glebius@FreeBSD.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <X11/keysym.h>

#include <sys/stat.h>
#include <sys/time.h>
#include <unistd.h>
#include <errno.h>
#include <fcntl.h>

#include <xf86.h>
#include <xf86Xinput.h>
#include <xf86_OSproc.h>
#include <exevents.h>
#include <xorgVersion.h>
#include <xkbsrv.h>

static XF86ModuleVersionInfo eGalaxVersionRec =
{
    "egalax",
    MODULEVENDORSTRING,
    MODINFOSTRING1,
    MODINFOSTRING2,
    XORG_VERSION_CURRENT,
    PACKAGE_VERSION_MAJOR, PACKAGE_VERSION_MINOR,
    PACKAGE_VERSION_PATCHLEVEL,
    ABI_CLASS_XINPUT,
    ABI_XINPUT_VERSION,
    MOD_CLASS_XINPUT,
    {0, 0, 0, 0}
};

static void eGalaxUnplug(pointer p);
static pointer eGalaxPlug(pointer, pointer, int *, int *);
static int eGalaxPreInit(InputDriverPtr, InputInfoPtr, int);
static void eGalaxUnInit(InputDriverPtr, InputInfoPtr, int);
static void eGalaxReadInput(InputInfoPtr);
static int eGalaxControl(DeviceIntPtr, int);
static void eGalaxCtrl(DeviceIntPtr, PtrCtrl *);
static void eGalaxInitAxes(DeviceIntPtr);
static void eGalaxConfigAxes(DeviceIntPtr);
static void eGalaxCalib(InputInfoPtr, int, int);

_X_EXPORT InputDriverRec EGALAX = {
	1,			/* driver version */
	"egalax",		/* driver name */
	NULL,			/* identify */
	eGalaxPreInit,		/* pre-init */
	eGalaxUnInit,		/* un-init */
	NULL,			/* module */
	0			/* ref count */
};

_X_EXPORT XF86ModuleData egalaxModuleData =
{
	&eGalaxVersionRec,
	eGalaxPlug,
	eGalaxUnplug
};

struct eGalaxDeviceRec {
	char		*device;
	u_short		minx, maxx, miny, maxy;
	u_char		revy;
	u_char		swapaxes;

	/* Second button emulator. */
	u_char		button, area, pause;
	u_short		ox, oy;
	struct timeval	tv;
};

typedef struct eGalaxDeviceRec * eGalaxDevicePtr;

static void
eGalaxUnplug(pointer p)
{
}

static pointer
eGalaxPlug(pointer module, pointer options, int *errmaj, int *errmin)
{
	xf86AddInputDriver(&EGALAX, module, 0);
	return (module);
}

static int
eGalaxPreInit(InputDriverPtr drv, InputInfoPtr pInfo, int flags)
{
	eGalaxDevicePtr sc;

	if ((sc = calloc(1, sizeof(struct eGalaxDeviceRec))) == NULL)
		return (BadAlloc);

	pInfo->type_name = XI_MOUSE; /* see XI.h */
	pInfo->device_control = eGalaxControl; /* enable/disable dev */
	pInfo->read_input = eGalaxReadInput; /* new data avl */
	pInfo->control_proc = NULL;
	pInfo->switch_mode = NULL;
	pInfo->dev = NULL;
	pInfo->private = sc;

	/* process driver specific options */
	sc->device = xf86SetStrOption(pInfo->options, "Device", "/dev/uep0");
	xf86Msg(X_INFO, "%s: Using device %s.\n", pInfo->name, sc->device);

	/* process generic options */
	xf86CollectInputOptions(pInfo, NULL);
	xf86ProcessCommonOptions(pInfo, pInfo->options);

	/*
	 * Default values are initially much narrower than real,
	 * the screen will autocalibrate soon.
	 */
	sc->minx = xf86SetIntOption(pInfo->options, "MinX", 500);
	sc->maxx = xf86SetIntOption(pInfo->options, "MaxX", 1500);
	sc->miny = xf86SetIntOption(pInfo->options, "MinY", 500);
	sc->maxy = xf86SetIntOption(pInfo->options, "MaxY", 1500);

	sc->revy = xf86SetBoolOption(pInfo->options, "ReverseY", TRUE);
	sc->swapaxes = xf86SetBoolOption(pInfo->options, "SwapAxes", FALSE);

	sc->area = xf86SetIntOption(pInfo->options, "RightClickEmulArea", 5);
	sc->pause = xf86SetIntOption(pInfo->options, "RightClickEmulPause", 1);

	/* Open sockets, init device files, etc. */
	pInfo->fd = open(sc->device, O_RDWR | O_NONBLOCK);
	if (pInfo->fd == -1) {
		xf86Msg(X_ERROR, "%s: failed to open %s.", pInfo->name,
		    sc->device);
		pInfo->private = NULL;
		free(sc);
		xf86DeleteInput(pInfo, 0);
		return (BadValue);
	}
	close(pInfo->fd);
	pInfo->fd = -1;

	return (Success);
}

static void
eGalaxUnInit(InputDriverPtr drv, InputInfoPtr pInfo, int flags)
{
	eGalaxDevicePtr sc = pInfo->private;

	if (sc->device)
		free(sc->device);
	free(sc);
	xf86DeleteInput(pInfo, 0);
}

static int
eGalaxControl(DeviceIntPtr device, int what)
{
	InputInfoPtr pInfo = device->public.devicePrivate;
	eGalaxDevicePtr sc = pInfo->private;
	int ret = Success;

	switch (what) {
	case DEVICE_INIT:
		eGalaxInitAxes(device);
		break;

	/* Switch device on.  Establish socket, start event delivery.  */
	case DEVICE_ON:
		xf86Msg(X_INFO, "%s: On.\n", pInfo->name);
		if (device->public.on)
			break;

		pInfo->fd = open(sc->device, O_RDONLY | O_NONBLOCK);
		if (pInfo->fd < 0) {
			xf86Msg(X_ERROR, "%s: cannot open device.\n",
			    pInfo->name);
			return (BadRequest);
		}

		xf86FlushInput(pInfo->fd);
		xf86AddEnabledDevice(pInfo);
		device->public.on = TRUE;
		break;

	case DEVICE_OFF:
		xf86Msg(X_INFO, "%s: Off.\n", pInfo->name);
		if (!device->public.on)
			break;
		xf86RemoveEnabledDevice(pInfo);
		close(pInfo->fd);
		pInfo->fd = -1;
		device->public.on = FALSE;
		break;

	case DEVICE_CLOSE:
		/* free what we have to free */
		break;
	}

	return (ret);
}

static void
eGalaxReadInput(InputInfoPtr pInfo)
{
	eGalaxDevicePtr sc = pInfo->private;
	char buf[5];
	int x, y, nread, reinit;

	while(xf86WaitForInput(pInfo->fd, 0) > 0) {
		nread = read(pInfo->fd, &buf, 5);
		if (nread <= 0) {
			if (errno == ENXIO) {
				xf86Msg(X_ERROR, "%s: Device disappeared\n",
				    pInfo->name);
				xf86RemoveEnabledDevice(pInfo);
				close(pInfo->fd);
				pInfo->fd = -1;
			} else if (errno != EAGAIN)
				xf86Msg(X_ERROR, "%s: Read error: %s\n",
				    pInfo->name, strerror(errno));
			return;
		}
		if (nread < 5) {
			xf86Msg(X_ERROR, "%s: bad packet len %u.\n",
			    pInfo->name, nread);
			return;
		}

		x = (buf[1] << 7) | buf[2];
		y = (buf[3] << 7) | buf[4];

		if (sc->swapaxes) {
			int tmp;

			tmp = x;
			x = y;
			y = tmp;
		}

		if (x < sc->minx || x > sc->maxx ||
		    y < sc->miny || y > sc->maxy)
			eGalaxCalib(pInfo, x, y);

		if (sc->revy)
			y = sc->maxy - y + sc->miny;

		xf86PostMotionEvent(pInfo->dev,
		    TRUE, /* is_absolute */
		    0, /* first_valuator */
		    2, /* num_valuators */
		    x, y);

		/* Touchpad is released. */
		if ((buf[0] & 0x01) == 0 && sc->button == 0)
			continue;

		if ((buf[0] & 0x01) == 0 && sc->button > 0) {
			/* Release button. */
			xf86PostButtonEvent(pInfo->dev, TRUE, sc->button, 0,
			    0, 2, x, y);
			sc->button = 0;
			timerclear(&sc->tv);
			continue;
		}

		/* Touchpad is pressed. */
		if ((buf[0] & 0x01) == 0x01 && sc->button == 0) {
			struct timeval tv;

			xf86PostButtonEvent(pInfo->dev, TRUE, 1, 1, 0, 2, x, y);
			sc->button = 1;

			gettimeofday(&sc->tv, NULL);
			sc->ox = x;
			sc->oy = y;
			continue;
		}
		

		if ((buf[0] & 0x01) == 0x01 && sc->button == 1 &&
		    (abs(x - sc->ox) < sc->area ) &&
		    (abs(y - sc->oy) < sc->area )) {
			struct timeval tv, res;

			gettimeofday(&tv, NULL);
			timersub(&tv, &sc->tv, &res);
			if (res.tv_sec >= sc->pause) {
				xf86PostButtonEvent(pInfo->dev, TRUE, 1, 0,
				    0, 2, x, y);
				xf86PostButtonEvent(pInfo->dev, TRUE, 3, 1,
				    0, 2, x, y);
				sc->button = 3;
			}
		}

	} /* while */
}

static void
eGalaxCtrl(DeviceIntPtr device, PtrCtrl *ctrl)
{
}

static void
eGalaxInitAxes(DeviceIntPtr device)
{
	InputInfoPtr pInfo = device->public.devicePrivate;
	eGalaxDevicePtr sc = pInfo->private;
	unsigned char map[] = {0, 1, 2, 3};
	Atom btn_label;
	Atom axis_labels[2] = { 0, 0 };

	InitPointerDeviceStruct((DevicePtr)device, map, 3, &btn_label,
	    eGalaxCtrl, GetMotionHistorySize(), 2, axis_labels);

	eGalaxConfigAxes(device);
}

static void
eGalaxConfigAxes(DeviceIntPtr device)
{
	InputInfoPtr pInfo = device->public.devicePrivate;
	eGalaxDevicePtr sc = pInfo->private;
	Atom axis_labels[2] = { 0, 0 };

	xf86InitValuatorAxisStruct(device, 0, axis_labels[0],
	    sc->minx, sc->maxx, 1, 1, 1, Absolute);
	xf86InitValuatorDefaults(device, 0);

	xf86InitValuatorAxisStruct(device, 1, axis_labels[1],
	    sc->miny, sc->maxy, 1, 1, 1, Absolute);
	xf86InitValuatorDefaults(device, 1);
}

static void
eGalaxCalib(InputInfoPtr pInfo, int x, int y)
{
	eGalaxDevicePtr sc = pInfo->private;

	if (x < sc->minx)
		sc->minx = x;

	if (x > sc->maxx)
		sc->maxx = x;

	if (y < sc->miny)
		sc->miny = y;

	if (y > sc->maxy)
		sc->maxy = y;

	eGalaxConfigAxes(pInfo->dev);

	xf86Msg(X_WARNING,
	    "%s: adjusted calibration MinX=%u, MaxX=%u, MinY=%u, MaxY=%u.\n",
	    pInfo->name, sc->minx, sc->maxx, sc->miny, sc->maxy);
}
